internal/provider/project_access_resource.go

```diff
@@ -0,0 +1,276 @@
+}
+
+func (r *projectAccessResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	tflog.Debug(ctx, "Preparing to create projectAccess resource")
+	var plan projectAccessResourceModel
+
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	tflog.Info(ctx, fmt.Sprintf("Upserting %v", plan))
+	resp.Diagnostics.Append(r.upsertProjectAccess(ctx, plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, &plan)...)
+	tflog.Debug(ctx, "Finished creating projectAccess data source", map[string]any{"success": true})
+}
+
+func (r *projectAccessResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	tflog.Debug(ctx, "Preparing to read projectAccess resource")
+	var state projectAccessResourceModel
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	projectId := *state.Project.ValueStringPointer()
+
+	projectAccess, api_response, err := r.client.ProjectsAPI.GetProjectAccess(ctx, projectId).Execute()
+
+	if err != nil {
+		resp.Diagnostics.AddError(
+			fmt.Sprintf("Unable to Read ProjectAccess %s", state.Project.ValueString()),
+			err.Error(),
+		)
+		return
+	}
+
+	if api_response.StatusCode != 200 {
+		resp.Diagnostics.AddError(
+			"Unexpected HTTP error code received",
+			api_response.Status,
+		)
+		return
+	}
+
+	state.Roles = transformToInternalRoles(projectAccess)
+
+	tflog.Info(ctx, fmt.Sprintf("Read projectAccess %v", state))
+	resp.Diagnostics.Append(resp.State.Set(ctx, &state)...)
+	tflog.Debug(ctx, "Finished reading projectAccess data source", map[string]any{"success": true})
+}
+
+func (r *projectAccessResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
+	tflog.Debug(ctx, "Preparing to update project access resource")
+	var plan projectAccessResourceModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	resp.Diagnostics.Append(r.upsertProjectAccess(ctx, plan)...)
+
+	var state projectAccessResource
+	req.State.Get(ctx, &state)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, &plan)...)
+	tflog.Debug(ctx, "Finished updating projectAccess data source", map[string]any{"success": true})
+}
+
+func (r *projectAccessResource) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
+	resp.Diagnostics.AddWarning("Resource Not Deleted", "The projectAccess resource was removed from the Terraform state, but not deleted from the actual system. This is to avoid potential mistakes. Instead of deleting projectAccess you may just delete the whole project")
+}
+
+/** Helper methods **/
+
+func (r *projectAccessResource) upsertProjectAccess(ctx context.Context, plan projectAccessResourceModel) diag.Diagnostics {
```

Comment:
This one is scary! Don't think we can do anything about it now but there's a lot of things that can go wrong here and backtracking state when something goes belly up is going to be hard!

We can't do this on the unleash side as a bulk operation?

Reply:
It is, we just send the desired state and then Unleash server handles it in a transactional way: https://github.com/ivarconr/unleash-enterprise/blob/cb178bc30d1232ad5f3c83473e45960327d7af93/src/project-access/project-access-service.ts#L32-L48

Reply:
I'm dumb, sorry, I realise this is one call just a large method